database/src/service/tests.rs

```diff
@@ -7,76 +7,364 @@
 
 // This is only imported on `#[cfg(test)]` in `mod.rs`.
 
-#![allow(unused_mut, clippy::significant_drop_tightening)]
+#![allow(
+    clippy::significant_drop_tightening,
+    clippy::await_holding_lock,
+    clippy::too_many_lines
+)]
 
 //---------------------------------------------------------------------------------------------------- Use
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+
+use pretty_assertions::assert_eq;
 use tower::{Service, ServiceExt};
 
-use cuprate_types::service::{ReadRequest, Response, WriteRequest};
+use cuprate_test_utils::data::{block_v16_tx0, block_v1_tx2, block_v9_tx3};
+use cuprate_types::{
+    service::{ReadRequest, Response, WriteRequest},
+    ExtendedBlockHeader, OutputOnChain, VerifiedBlockInformation,
+};
 
 use crate::{
     config::Config,
+    free::output_to_output_on_chain,
+    ops::{
+        block::{get_block_extended_header_from_height, get_block_info},
+        blockchain::top_block_height,
+        output::get_output,
+    },
     service::{init, DatabaseReadHandle, DatabaseWriteHandle},
+    tables::{KeyImages, Tables, TablesIter},
+    tests::AssertTableLen,
+    types::{Amount, AmountIndex, KeyImage, PreRctOutputId},
+    ConcreteEnv, DatabaseIter, DatabaseRo, Env, EnvInner, RuntimeError,
 };
 
-//---------------------------------------------------------------------------------------------------- Tests
+//---------------------------------------------------------------------------------------------------- Helper functions
 /// Initialize the `service`.
-fn init_service() -> (DatabaseReadHandle, DatabaseWriteHandle, tempfile::TempDir) {
+fn init_service() -> (
+    DatabaseReadHandle,
+    DatabaseWriteHandle,
+    Arc<ConcreteEnv>,
+    tempfile::TempDir,
+) {
     let tempdir = tempfile::tempdir().unwrap();
     let config = Config::low_power(Some(tempdir.path().into()));
     let (reader, writer) = init(config).unwrap();
-    (reader, writer, tempdir)
+    let env = reader.env().clone();
+    (reader, writer, env, tempdir)
+}
+
+/// This is the template used in the actual test functions below.
+///
+/// - Send write request(s)
+/// - Receive response(s)
+/// - Assert proper tables were mutated
+/// - Assert read requests lead to expected responses
+#[allow(clippy::future_not_send)] // INVARIANT: tests are using a single threaded runtime
+async fn test_template(
+    // Which block(s) to add?
+    block_fns: &[fn() -> &'static VerifiedBlockInformation],
+    // Total amount of generated coins after the block(s) have been added.
+    cumulative_generated_coins: u64,
+    // What are the table lengths be after the block(s) have been added?
+    assert_table_len: AssertTableLen,
+) {
```

Comment:
Tests all `Request`'s lead to expected `Response`'s, which is currently tested on...